### PR TITLE
Adding `#pragma CPROVER enable "named-check"`

### PR DESCRIPTION
**Pull request by remi-delmas-3000**
This PR adds the possibility to selectively activate checks :
- on blocks of C code using `#pragma CPROVER enable "named-check"`
- to individual GOTO instructions and expressions using `source_location.add_pragma("enable:named-check")`

It is possible to enable/disable a check, to push a level on the stack and override the check status using another pragma enable/disable pragma.
Attempts to enable and disable a same check at the same stack level are detected and cause a PARSING_ERROR in the C front end and a INVARIANT violation in `goto_check`.

Modifications:
- the collection of allocations on the whole goto program using `goto_checkt::collect_allocation` is now always on
- a `messaget log` attribute is added to `goto_checkt` to be able to log warnings in case a same check is enabled/disabled more than once on a given GOTO instruction
- the `flag_resett` class is extended to detect multiple 'set_flag' operations on a same flag
- the ansi-c lexer/parser is modified to accept `#pragma CPROVER enable "named check"` and flag shadowing
- tests are added in `regression/cbmc/pragma_cprover_enable*` to test the feature

This feature is needed as part of the function and loop contracts verification, to be able to selectively activate checks on  instructions used for assigns-clause checking, in order to guarantee the soundness of the analysis, even if these checks are not globally activated in the current or ulterior `goto-instrument` or `cbmc` passes

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/analyses/goto_check.cpp [hunk 1]
```diff
@@ -99,6 +103,12 @@ class goto_checkt
   std::unique_ptr<local_bitvector_analysist> local_bitvector_analysis;
   goto_programt::const_targett current_target;
   guard_managert guard_manager;
+  messaget log;
+
+  /// when true causes a forced collection of all allocations
+  /// (needed to allow selectively enabling checks per instruction)
+  bool do_collect_allocations;
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: It seems like we are setting this to `true` in the constructor and no other method sets it to `false`, so wouldn't this be _always_ `true`? And if that's the case, can we completely remove this and also remove the `if(!... & !...) return;` check in `collect_allocations`?

Or did I miss a code path where it won't be `true`?

File: src/analyses/goto_check.cpp [hunk 2]
```diff
@@ -1856,24 +1866,47 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
+  }
+
   /// Store the current value of \p flag and then set its value to \p new_value.
   void set_flag(bool &flag, bool new_value)
   {
+    if(warn_on_seen && seen(flag))
+    {
+      log.warning() << "Found more than one CPROVER enable/disable check "
+                       "pragma on a same goto instruction"
+                    << messaget::eom;
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: I think the error message could be improved a bit. Having "more than one CPROVER enable/disable check" is fine as long as they are _different_ checks.
I think we want to say that we _the same check_ being enabled/disabled multiple times times at a goto instruction.

File: regression/cbmc/pragma_cprover_enable3/test.desc [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+^\[main.pointer_primitives.\d+\] line 11 pointer invalid in POINTER_OBJECT\(\(const void \*\)q\): FAILURE$
+^\[main.pointer_primitives.\d+\] line 11 deallocated dynamic object in POINTER_OBJECT\(\(const void \*\)q\): SUCCESS$
+^\[main.pointer_primitives.\d+\] line 11 dead object in POINTER_OBJECT\(\(const void \*\)q\): SUCCESS$
+^\[main.pointer_primitives.\d+\] line 11 pointer outside object bounds in POINTER_OBJECT\(\(const void \*\)q\): FAILURE$
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+--
+^\[main.pointer_primitives.\d+\] line 17 
+--
```
Comment (SaswatPadhi) [inline regression/cbmc/pragma_cprover_enable3/test.desc]: [nit] missing newline. Also in a few other files.

File: regression/cbmc/pragma_cprover_enable1/main.c [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+{
+  int x;
+  int y[1];
+
+#pragma CPROVER check push
+#pragma CPROVER check enable "bounds"
+#pragma CPROVER check enable "signed-overflow"
+  // generate assertions for the following statement
+  x = x + y[1];
+  // pop all annotations
```
Comment (SaswatPadhi) [inline regression/cbmc/pragma_cprover_enable1/main.c]:
```suggestion
  // pop all checks added since the last push
```

[nit] also in other test files

**Review by SaswatPadhi — COMMENTED** [review]
Nice :) Mostly minor comments below.

Can we add at least one regression test to check nesting of both disable and enable pragmas, something like:

```c
  int n;

#pragma CPROVER check push
#pragma CPROVER check enable "signed-overflow"
   // generate overflow checks for the following addition
  int x = n + n;
#pragma CPROVER check push
#pragma CPROVER check disable "signed-overflow"
   // do not generate overflow checks for the following increment operations
  ++n;
  n++;
#pragma CPROVER check pop
   // generate overflow checks for the following addition
  int b = n + n;
#pragma CPROVER check pop
```

File: doc/cprover-manual/properties.md [hunk 5]
```diff
@@ -134,10 +134,13 @@ As all of these checks apply across the entire input program, we may wish to
...
 overflows can be expected and acceptable in certain instructions even when
 elsewhere we do not expect them. As of version 5.12, CBMC supports selectively
-disabling automatically generated properties.  To disable property generation,
+disabling or enabling automatically generated properties. 
+
+To disable property generation,
 use `#pragma CPROVER check disable "<name_of_check>"`, which remains in effect
 until a `#pragma CPROVER check pop` (to re-enable all properties
-disabled before or since the last `#pragma CPROVER check push`) is provided.
+disabled or enabled before or since the last `#pragma CPROVER check push`) is provided.
```
Comment (feliperodri) [inline doc/cprover-manual/properties.md]:
```
(to re-enable all properties disabled or enabled before or since the last `#pragma CPROVER check push`)
```
This wording seems confusing. Could you re-write it?

File: doc/cprover-manual/properties.md [hunk 6]
```diff
@@ -149,6 +152,22 @@ unsigned foo(unsigned x)
   x = x + 2; // unsigned overflow checks are generated here
 ```
 
+To enable property generation,
+use `#pragma CPROVER check enable "<name_of_check>"`, which remains in effect
+until a `#pragma CPROVER check pop` (to re-enable all properties
+disabled or enabled before or since the last `#pragma CPROVER check push`) is provided.
```
Comment (feliperodri) [inline doc/cprover-manual/properties.md]: Same as above.

File: regression/cbmc/pragma_cprover_enable1/main.c (hunk 4 above)
Comment (feliperodri) [reply regression/cbmc/pragma_cprover_enable1/main.c]: +1

File: regression/cbmc/pragma_cprover_enable1/test.desc [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+^\[main\.array_bounds\.1\] line \d+ array 'y' upper bound.*FAILURE$
+^\[main\.overflow\.1\] line \d+ arithmetic overflow on signed.*FAILURE$
+^\*\* 2 of 2 failed
+^VERIFICATION FAILED$
+^EXIT=10$
+^SIGNAL=0$
+--
+--
+Checks that we can selectively activate checks using pragmas.
```
Comment (feliperodri) [inline regression/cbmc/pragma_cprover_enable1/test.desc]: Add newline.

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (feliperodri) [reply src/analyses/goto_check.cpp]: +1

**Review by feliperodri — APPROVED** [review]
Only minor comments (modulo Saswat's comments).

**Comment by chrisr-diffblue** [comment]
Just a general git housekeeping comment - it's generally better to use `git pull --rebase origin develop` (replace `origin` with whatever you've called your remote - `origin` being the default if you didn't specify/change the remote name) to update your branch with the latest `develop` branch, rather than doing a merge of `develop` - it will lead to cleaner commits and cleaner git history.

File: src/goto-instrument/goto_instrument_parse_options.cpp [hunk 8]
```diff
@@ -1332,7 +1332,7 @@ void goto_instrument_parse_optionst::instrument_goto_program()
   }
 
   // add generic checks, if needed
-  goto_check(options, goto_model);
```
Comment (tautschnig) [inline src/goto-instrument/goto_instrument_parse_options.cpp:1335]: Use `ui_message_handler` here (should be the same as `log.get_message_handler()`, but is more consistent with the other code over here.

File: src/ansi-c/scanner.l [hunk 9]
```diff
@@ -402,29 +403,33 @@ void ansi_c_scanner_init()
...
+                  {
+                    yyansi_cerror(
+                      "Found enable and disable pragmas for " +
+                      id2string(check_name));
+                    return TOK_SCANNER_ERROR;
+                  }
+                  PARSER.pragma_cprover_add_check(check_name, enable);
                   PARSER.set_pragma_cprover();
                 }
-
```
Comment (tautschnig) [inline src/ansi-c/scanner.l:427]: Unnecessary removal of blank line.

File: src/ansi-c/ansi_c_parser.h [hunk 10]
```diff
@@ -143,15 +151,25 @@ class ansi_c_parsert:public parsert
...
+  /// \brief Pushes an empty level in the CPROVER pragma stack
+  void pragma_cprover_push();
+
+  /// \brief Pops a level in the CPROVER pragma stack
+  void pragma_cprover_pop();
+
+  /// \brief Adds a check to the CPROVER pragma stack
+  void pragma_cprover_add_check(irep_idt name, bool enabled);
+
+  /// Returns true the same check with different polarity
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: "Returns true iff the same" (missing "iff")

File: src/ansi-c/ansi_c_parser.h [hunk 11]
```diff
@@ -117,20 +123,22 @@ class ansi_c_parsert:public parsert
...
+    const scopet &current = current_scope();
     scopes.push_back(scopet());
-    scopes.back().prefix=current.prefix+prefix;
+    scopes.back().prefix = current.prefix + prefix;
   }
 
   ansi_c_id_classt lookup(
     const irep_idt &base_name, // in
-    irep_idt &identifier, // out
+    irep_idt &identifier,      // out
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Unnecessary whitespace change?

File: src/ansi-c/ansi_c_parser.h [hunk 12]
```diff
@@ -117,20 +123,22 @@ class ansi_c_parsert:public parsert
...
     parse_tree.items.push_back(declaration);
   }
 
   void new_scope(const std::string &prefix)
   {
-    const scopet &current=current_scope();
+    const scopet &current = current_scope();
     scopes.push_back(scopet());
-    scopes.back().prefix=current.prefix+prefix;
+    scopes.back().prefix = current.prefix + prefix;
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h [hunk 13]
```diff
@@ -107,7 +107,13 @@ class ansi_c_parsert:public parsert
...
   }
 
-  enum class decl_typet { TAG, MEMBER, PARAMETER, OTHER };
+  enum class decl_typet
+  {
+    TAG,
+    MEMBER,
+    PARAMETER,
+    OTHER
+  };
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h [hunk 14]
```diff
@@ -48,12 +48,12 @@ class ansi_c_parsert:public parsert
     parse_tree.clear();
 
     // scanner state
-    tag_following=false;
-    asm_block_following=false;
-    parenthesis_counter=0;
+    tag_following = false;
+    asm_block_following = false;
+    parenthesis_counter = 0;
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h [hunk 15]
```diff
@@ -39,7 +39,7 @@ class ansi_c_parsert:public parsert
 
   virtual bool parse() override
   {
-    return yyansi_cparse()!=0;
+    return yyansi_cparse() != 0;
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.h [hunk 16]
```diff
@@ -6,21 +6,21 @@ Author: Daniel Kroening, kroening@kroening.com
...
 #include <util/config.h>
+#include <util/parser.h>
 
 #include "ansi_c_parse_tree.h"
 #include "ansi_c_scope.h"
 
 int yyansi_cparse();
 
-class ansi_c_parsert:public parsert
+class ansi_c_parsert : public parsert
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.h]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp [hunk 17]
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
...
+  if(type.id() == ID_typedef)
     return ansi_c_id_classt::ANSI_C_TYPEDEF;
-  else if(type.id()==ID_struct ||
-          type.id()==ID_union ||
-          type.id()==ID_c_enum)
+  else if(
+    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_c_enum)
     return ansi_c_id_classt::ANSI_C_TAG;
-  else if(type.id()==ID_merged_type)
+  else if(type.id() == ID_merged_type)
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp [hunk 18]
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
...
 ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
 {
-  if(type.id()==ID_typedef)
+  if(type.id() == ID_typedef)
     return ansi_c_id_classt::ANSI_C_TYPEDEF;
-  else if(type.id()==ID_struct ||
-          type.id()==ID_union ||
-          type.id()==ID_c_enum)
+  else if(
+    type.id() == ID_struct || type.id() == ID_union || type.id() == ID_c_enum)
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: src/ansi-c/ansi_c_parser.cpp [hunk 19]
```diff
@@ -156,13 +143,12 @@ void ansi_c_parsert::add_declarator(
 
 ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
 {
-  if(type.id()==ID_typedef)
+  if(type.id() == ID_typedef)
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace change?

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 20]
```diff
@@ -812,7 +812,11 @@ void jbmc_parse_optionst::process_goto_function(
 
   // add generic checks
   goto_check(
-    function.get_function_id(), function.get_goto_function(), ns, options);
+    function.get_function_id(),
+    function.get_goto_function(),
+    ns,
+    options,
+    log.get_message_handler());
```
Comment (tautschnig) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: Use `ui_message_handler`

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 21]
```diff
@@ -705,7 +705,11 @@ void janalyzer_parse_optionst::process_goto_function(
 
   // add generic checks
   goto_check(
-    function.get_function_id(), function.get_goto_function(), ns, options);
+    function.get_function_id(),
+    function.get_goto_function(),
+    ns,
+    options,
+    log.get_message_handler());
```
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Use `ui_message_handler`

File: jbmc/src/jdiff/jdiff_parse_options.cpp [hunk 22]
```diff
@@ -278,7 +278,7 @@ bool jdiff_parse_optionst::process_goto_program(
 
     // add generic checks
     log.status() << "Generic Property Instrumentation" << messaget::eom;
-    goto_check(options, goto_model);
+    goto_check(options, goto_model, log.get_message_handler());
```
Comment (tautschnig) [inline jbmc/src/jdiff/jdiff_parse_options.cpp]: Use `ui_message_handler`

File: src/analyses/goto_check.cpp [hunk 23]
```diff
@@ -1892,7 +1904,6 @@ void goto_checkt::goto_check(
     util_make_unique<local_bitvector_analysist>(goto_function, ns);
 
   goto_programt &goto_program=goto_function.body;
-
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Removal of blank line?

File: src/analyses/goto_check.cpp [hunk 24]
```diff
@@ -1902,32 +1913,64 @@ void goto_checkt::goto_check(
...
+        seen = flag_resetter.set_flag(enable_undefined_shift_check, true);
+      else if(d.first == "enable:nan-check")
+        seen = flag_resetter.set_flag(enable_nan_check, true);
+      else if(d.first == "enable:pointer-primitive-check")
+        seen = flag_resetter.set_flag(enable_pointer_primitive_check, true);
+
+      INVARIANT(
+        !seen,
+        "Found enable and disable pragmas for " + id2string(d.first) +
+          " at \n" + i.source_location().pretty());
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Shouldn't this invariant just be maintained within `set_flag`? It seems it is checked every time you use `set_flag`.

File: src/ansi-c/ansi_c_parser.cpp [hunk 25]
```diff
@@ -19,21 +19,18 @@ ansi_c_id_classt ansi_c_parsert::lookup(
   bool label)
 {
   // labels and tags have a separate name space
-  const irep_idt scope_name=
-    tag?"tag-"+id2string(base_name):
-    label?"label-"+id2string(base_name):
-    base_name;
+  const irep_idt scope_name =
+    tag ? "tag-" + id2string(base_name)
+        : label ? "label-" + id2string(base_name) : base_name;
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Whitespace changes?

File: src/ansi-c/ansi_c_parser.cpp [hunk 26]
```diff
@@ -19,21 +19,18 @@ ansi_c_id_classt ansi_c_parsert::lookup(
...
 
-  for(scopest::const_reverse_iterator it=scopes.rbegin();
-      it!=scopes.rend();
+  for(scopest::const_reverse_iterator it = scopes.rbegin(); it != scopes.rend();
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Please review all changes in this file for whitespace-only changes.

**Review by tautschnig — CHANGES_REQUESTED** [review]
Functionality looks ok, but I think this PR needs some cleanup to remove as much noise as possible.

File: src/analyses/goto_check.cpp (hunk 1 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: this variable was removed and collection is now always on

File: src/analyses/goto_check.cpp (hunk 2 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: message updated to include the name of the check

File: regression/cbmc/pragma_cprover_enable3/test.desc (hunk 3 above)
Comment (remi-delmas-3000) [reply regression/cbmc/pragma_cprover_enable3/test.desc]: fixed

File: regression/cbmc/pragma_cprover_enable1/main.c (hunk 4 above)
Comment (remi-delmas-3000) [reply regression/cbmc/pragma_cprover_enable1/main.c]: comment removed

File: doc/cprover-manual/properties.md (hunk 5 above)
Comment (remi-delmas-3000) [reply doc/cprover-manual/properties.md]: documentation updated

File: doc/cprover-manual/properties.md (hunk 6 above)
Comment (remi-delmas-3000) [reply doc/cprover-manual/properties.md]: documentation updated

File: jbmc/src/jbmc/jbmc_parse_options.cpp (hunk 20 above)
Comment (remi-delmas-3000) [reply jbmc/src/jbmc/jbmc_parse_options.cpp]: corrected

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 21 above)
Comment (remi-delmas-3000) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: corrected

File: jbmc/src/jdiff/jdiff_parse_options.cpp (hunk 22 above)
Comment (remi-delmas-3000) [reply jbmc/src/jdiff/jdiff_parse_options.cpp]: corrected

File: src/analyses/goto_check.cpp (hunk 23 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: corrected

File: src/analyses/goto_check.cpp (hunk 24 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: The invariant is now maintained inside `flag_resett`. To be able to generate readable error messages`flag_resett::set_flag` now takes the flag name and instruction as parameters.

File: src/goto-instrument/goto_instrument_parse_options.cpp (hunk 8 above)
Comment (remi-delmas-3000) [reply src/goto-instrument/goto_instrument_parse_options.cpp:1335]: corrected

File: src/ansi-c/ansi_c_parser.h [hunk 27]
```diff
@@ -10,6 +10,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #ifndef CPROVER_ANSI_C_ANSI_C_PARSER_H
 #define CPROVER_ANSI_C_ANSI_C_PARSER_H
 
+#include <map>
```
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.h:13]: Do we still need the `#include <set>` below?

File: src/ansi-c/ansi_c_parser.h [hunk 28]
```diff
@@ -143,15 +144,25 @@ class ansi_c_parsert:public parsert
...
-  {
-    source_location.remove(ID_pragma);
-    for(const auto &pragma_set : pragma_cprover)
-    {
-      for(const auto &pragma : pragma_set)
-        source_location.add_pragma(pragma);
-    }
-  }
+  /// \brief True iff the CPROVER pragma stack is empty
+  bool pragma_cprover_empty();
```
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.h:146]: I see why you have introduced these functions although `pragma_cprover_stack` is `public`.

Can we make it `private` / `protected` now? Just so we don't allow messing with `pragma_cprover_stack` directly and use these functions instead?

File: src/ansi-c/ansi_c_parser.cpp [hunk 29]
```diff
@@ -175,3 +175,56 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+
+  for(const auto &pragma_set : pragma_cprover_stack)
+    for(const auto &pragma : pragma_set)
+      flattened[pragma.first] = pragma.second;
+
+  source_location.remove(ID_pragma);
+
+  for(const auto &pragma : flattened)
+  {
+    std::string check_name = id2string(pragma.first);
```
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp:225]:
```suggestion
    auto check_name = id2string(pragma.first);
```

File: src/ansi-c/ansi_c_parser.cpp [hunk 30]
```diff
@@ -175,3 +175,56 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+    std::string full_name = pragma.second
+                              ? std::string("enable:") + check_name
+                              : std::string("disable:") + check_name;
```
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp]: Since `check_name` gets `std::string` (not `const char *`) type from the `auto` definition above, we can use the overloaded `+` in C++ to simplify this expression:

```suggestion
    auto full_name = (pragma.second ? "enable:" : "disable:") + check_name;
```

File: src/analyses/goto_check.cpp [hunk 31]
```diff
@@ -1902,32 +1917,62 @@ void goto_checkt::goto_check(
     const auto &pragmas = i.source_location().get_pragmas();
     for(const auto &d : pragmas)
     {
+      auto err_msg = [&]() {
+        return "Found enable and disable pragmas for " + id2string(d.first) +
+               " at \n" + i.source_location().pretty();
+      };
       if(d.first == "disable:bounds-check")
-        flag_resetter.set_flag(enable_bounds_check, false);
+        flag_resetter.set_flag(enable_bounds_check, false, err_msg);
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: I am not sure if I understand why `err_msg` needs to be a lambda function. It seems like it's called immediately (at the `INVARIANT` check) within the `set_flag` call. Can we not just pass a `string` here instead?

File: src/analyses/goto_check.cpp [hunk 32]
```diff
@@ -1995,7 +2040,11 @@ void goto_checkt::goto_check(
       // safety
       {
         flag_resett no_enum_check_flag_resetter;
-        no_enum_check_flag_resetter.set_flag(no_enum_check, true);
+        auto err_msg = [&]() {
+          return "no_enum_check flag disabled and enabled at \n" +
+                 i.source_location().pretty();
+        };
+        no_enum_check_flag_resetter.set_flag(no_enum_check, true, err_msg);
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Same comment as above.
Can we not just pass a `string` here instead?

File: src/analyses/goto_check.cpp [hunk 33]
```diff
@@ -1853,27 +1851,44 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
     {
       flags_to_reset.emplace_back(&flag, flag);
       flag = new_value;
     }
+    return;
   }
 
-  /// Restore the values of all flags that have been modified via `set_flag`.
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool &flag)
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Hmm, does this function require a mutable reference? I can see why `set_flag` does, but I don't think `seen` does? How about:

```suggestion
  bool seen(const bool &flag)
```

File: src/analyses/goto_check.cpp [hunk 34]
```diff
@@ -1853,27 +1851,44 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
+  {
+    return seen_flags.find(&flag) != seen_flags.end();
```
Comment (SaswatPadhi) [inline src/analyses/goto_check.cpp]: Hmm, we seem to be taking the address of `flag` here. So if we have two `flag`s with the same content but different addresses then this would break?

Do we have some mechanism to guard against it? Or if there's some other reason as to why this is a non-issue, we should mention that in the comments above this function.

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
A few more comments; nothing major

File: src/analyses/goto_check.cpp (hunk 34 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: the class `flag_resett` is file-local so as long as the discipline is respected in this file it should be ok.

File: src/ansi-c/ansi_c_parser.cpp (hunk 30 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: modif added

File: src/analyses/goto_check.cpp (hunk 31 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: this lambda is used to defer the creation of the error string until it is actually needed (if ever).

File: src/analyses/goto_check.cpp (hunk 32 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp:2048]: This lambda allows to defer the creation of the string until it is needed

File: src/analyses/goto_check.cpp [hunk 35]
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
+    INVARIANT(
+      !(seen(&flag) && flag != new_value),
+      "Flag " + id2string(flag_name) +
+        " set twice with incompatible values "
+        " at \n" +
+        instruction.source_location().pretty());
     if(flag != new_value)
     {
-      flags_to_reset.emplace_back(&flag, flag);
+      flags_to_reset.insert(std::pair<bool *, bool>(&flag, flag));
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: I might be wrong, but I believe you can shorten this to `flags_to_reset.insert({&flag, flag});` with C++11.

File: src/analyses/goto_check.cpp [hunk 36]
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
       flag = new_value;
     }
+    return;
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Seems redundant?

File: src/analyses/goto_check.cpp [hunk 37]
```diff
@@ -1851,30 +1849,54 @@ optionalt<exprt> goto_checkt::rw_ok_check(exprt expr)
...
   ~flag_resett()
   {
     for(const auto &flag_pair : flags_to_reset)
       *flag_pair.first = flag_pair.second;
   }
 
 private:
-  std::list<std::pair<bool *, bool>> flags_to_reset;
+  /// \brief True iff flag has already been seen by \ref set_flag
+  bool seen(bool *flag)
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Not that it matters a whole lot, but the parameter type should be `const bool *` to make clear that no modification will occur. Also, the method should be marked `const`.

File: src/analyses/goto_check.cpp [hunk 38]
```diff
@@ -1898,36 +1920,62 @@ void goto_checkt::goto_check(
     current_target = it;
     goto_programt::instructiont &i=*it;
 
-    flag_resett flag_resetter;
+    flag_resett resetter;
```
Comment (tautschnig) [inline src/analyses/goto_check.cpp]: Given that every call requires the instruction to be passed (and the instruction is always the same), should the constructor already take that?

File: src/ansi-c/ansi_c_parser.cpp [hunk 39]
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+
+bool ansi_c_parsert::pragma_cprover_empty()
+{
+  return pragma_cprover_stack.empty();
+}
+
+void ansi_c_parsert::pragma_cprover_push()
+{
+  std::map<irep_idt, bool> empty;
+  pragma_cprover_stack.push_back(std::move(empty));
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: I'm wondering whether `pragma_cprover_stack.emplace_back({});` would actually work (and be more concise)?

File: src/ansi-c/ansi_c_parser.cpp [hunk 40]
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+}
+
+void ansi_c_parsert::pragma_cprover_pop()
+{
+  pragma_cprover_stack.pop_back();
+}
+
+void ansi_c_parsert::pragma_cprover_add_check(irep_idt name, bool enabled)
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: Use `const irep_idt &`

File: src/ansi-c/ansi_c_parser.cpp [hunk 41]
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+{
+  if(pragma_cprover_stack.empty())
+    pragma_cprover_push();
+
+  pragma_cprover_stack.back()[name] = enabled;
+}
+
+bool ansi_c_parsert::pragma_cprover_clash(irep_idt name, bool enabled)
```
Comment (tautschnig) [inline src/ansi-c/ansi_c_parser.cpp]: `const irep_idt &`

**Review by tautschnig — APPROVED** [review]

File: src/analyses/goto_check.cpp (hunk 32 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: now passing flag name and instructions to set_flag

File: src/analyses/goto_check.cpp (hunk 35 above)
Comment (SaswatPadhi) [reply src/analyses/goto_check.cpp]: Or just `flags_to_reset.emplace(&flag, flag);` to avoid creating a temporary pair.

File: src/ansi-c/ansi_c_parser.cpp [hunk 42]
```diff
@@ -175,3 +175,55 @@ ansi_c_id_classt ansi_c_parsert::get_class(const typet &type)
...
+    for(const auto &pragma : pragma_set)
+      flattened[pragma.first] = pragma.second;
+
+  source_location.remove(ID_pragma);
+
+  for(const auto &pragma : flattened)
+  {
+    std::string check_name = id2string(pragma.first);
+    std::string full_name =
+      (pragma.second ? "enable:" : "disable:") + check_name;
```
Comment (SaswatPadhi) [inline src/ansi-c/ansi_c_parser.cpp:227]:
```suggestion
    auto check_name = id2string(pragma.first);
    auto full_name =
      (pragma.second ? "enable:" : "disable:") + check_name;
```

**Review by SaswatPadhi — APPROVED** [review]
Approving modulo Michael's remaining comments and one minor suggestion:

File: src/ansi-c/ansi_c_parser.cpp (hunk 42 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp:227]: `scanner.l ` uses `std::string` everywhere so I'll keep this as is for consistency

File: src/analyses/goto_check.cpp (hunk 36 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp (hunk 42 above)
Comment (SaswatPadhi) [reply src/ansi-c/ansi_c_parser.cpp:227]: Yes, but not sure why it has to be kept this way here though ... Using `auto` is not only more succinct but also makes it future proof (disconnects this from the underlying type, which is base `std::string` today)

File: src/analyses/goto_check.cpp (hunk 38 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp (hunk 39 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: seems to work with `pragma_cprover_stack.emplace_back()`

File: src/analyses/goto_check.cpp (hunk 37 above)
Comment (remi-delmas-3000) [reply src/analyses/goto_check.cpp]: method removed

File: src/ansi-c/ansi_c_parser.cpp (hunk 40 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp (hunk 41 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp]: fixed

File: src/ansi-c/ansi_c_parser.cpp (hunk 42 above)
Comment (remi-delmas-3000) [reply src/ansi-c/ansi_c_parser.cpp:227]: this would only be patching two lines out of several dozen that could use `auto`  the whole lexer would need refactoring to be effectively future proof.

**Comment by feliperodri** [comment]
@peterschrammel @martin-cs could you take a look at this PR? It's missing code owner approval.

File: src/ansi-c/ansi_c_parser.cpp (hunk 42 above)
Comment (SaswatPadhi) [reply src/ansi-c/ansi_c_parser.cpp:227]: Up to you 🙂

**Comment by martin-cs** [comment]
@feliperodri Apologies for the lag; I missed this one.

**Comment by martin-cs** [comment]
This would be easier to review if it were split up into multiple commits.  Parsing, implementation, testing would seem a reasonable start.

**Comment by remi-delmas-3000** [comment]
@martin-cs @feliperodri @tautschnig I reworked the commit history it should be good for the final review now

**Review by peterschrammel — APPROVED** [review]

**Comment by martin-cs** [comment]
Thank you @remi-delmas-3000 it makes reviewing so much easier!
